connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java

```diff
@@ -0,0 +1,105 @@
+     * executed exactly once.  If {@code timeoutDuration} is less than {@code retryBackoffMs}, the callable will be
+     * executed only once.
+     *
+     * <p>If {@code retryBackoffMs} is set to 0, no wait will happen in between the retries.
+     *
+     * @param callable          the function to execute.
+     * @param description       supplier that provides custom message for logging purpose
+     * @param timeoutDuration   timeout duration; may not be null
+     * @param retryBackoffMs    the number of milliseconds to delay upon receiving a
+     *                          {@link org.apache.kafka.connect.errors.RetriableException} before retrying again;
+     *                          must be 0 or more
+     * @throws ConnectException If the task exhausted all the retries.
+     */
+    public static <T> T retryUntilTimeout(Callable<T> callable, Supplier<String> description, Duration timeoutDuration, long retryBackoffMs) throws Exception {
+        if (timeoutDuration == null) {
+            throw new IllegalArgumentException("timeoutDuration cannot be null");
+        }
+
+        // if null supplier or string is provided, the message will be default to "callabe"
+        String descriptionStr = Optional.ofNullable(description)
+                .map(Supplier::get)
+                .orElse("callable");
+
+        long timeoutMs = timeoutDuration.toMillis();
+
+        if (retryBackoffMs >= timeoutMs) {
+            log.warn("Call to {} will only execute once, since retryBackoffMs={} is larger than total timeoutMs={}",
+                    descriptionStr, retryBackoffMs, timeoutMs);
+        }
+
+        if (retryBackoffMs >= timeoutMs ||
+                timeoutMs <= 0) {
+            // no retry
+            return callable.call();
+        }
+
+        long end = System.currentTimeMillis() + timeoutMs;
+        int attempt = 0;
+        Throwable lastError = null;
+        while (System.currentTimeMillis() <= end) {
+            attempt++;
+            try {
+                return callable.call();
+            } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
+                log.warn("Attempt {} to execute {} resulted in RetriableException; retrying automatically. " +
+                        "Reason: {}", attempt, descriptionStr, e.getMessage(), e);
+                lastError = e;
+            } catch (WakeupException e) {
+                lastError = e;
+            }
+
+            // if current time is less than the ending time, no more retry is necessary
+            // won't sleep if retryBackoffMs equals to 0
+            long millisRemaining = Math.max(0, end - System.currentTimeMillis());
+            if (millisRemaining > 0) {
+                Utils.sleep(retryBackoffMs);
+            }
+        }
+
+        throw new ConnectException("Fail to execute " + descriptionStr + " after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```

Comment:
Nit: including "execute" here is completely unnecessary.
```suggestion
        throw new ConnectException("Fail to " + descriptionStr + " after " + attempt + " attempts.  Reason: " + lastError.getMessage(), lastError);
```

Reply:
A comment on the logging message, I think they can be grammatically strange if the provided description isn't a verb. 
 E.g., "Fail to logging error". Probably not a big deal.